File: c++/longest-repeating-character-replacement.cc
class Solution {
public:
    int characterReplacement(string s, int k) {
        unordered_map<char, int> hashmap;
        int l = 0;
        int maxLen = 0;
        int maxFreq = 0;

        for (int r = 0; r < s.size(); r++) {
            hashmap[s[r]]++;
            maxFreq = max(maxFreq, hashmap[s[r]]);
            while (r - l + 1 - maxFreq > k) {
                hashmap[s[l]]--;
                maxFreq = max(maxFreq, hashmap[s[l]]);
                l++;
            }
            maxLen = max(maxLen, r - l + 1);            
        }
        return maxLen;
    }
};

File: python/longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        freqs = defaultdict(int)
        longest = 0
        i = 0

        for j in range(len(s)):
            freqs[s[j]] += 1
            maxFreq = max(freqs.values())
            curLen = j - i + 1
            if curLen - maxFreq > k:
                freqs[s[i]] -= 1
                i += 1
            longest = max(longest, j - i + 1)

        return longest
